### ISSUE TITLE
Various layout issues with tab bar controller

### ISSUE THREAD

**Issue by hyouuu** [MessageKit/MessageKit#85]
2 major issues - the overall layout is weird, and the sender name & date label alignment is ignored:
![simulator screen shot aug 26 2017 1 25 35 pm](https://user-images.githubusercontent.com/5061867/29744846-10c590fa-8a62-11e7-9427-dd58bc5eb551.png)

App Delegate looks like this to setup the tab bar controller:
```
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {

        window = UIWindow(frame: UIScreen.main.bounds)

        let iconSide: CGFloat = 34.0
        let iconSize = CGSize(width: iconSide, height: iconSide)
        let iconInsets = UIEdgeInsets(top: 6, left: 0, bottom: -6, right: 0)

        let home = StoryVC()
        home.view.backgroundColor = UIColor.orange
        home.tabBarItem.image = UIImage(named: "home")?.resizedImage(iconSize)
        home.tabBarItem.imageInsets = iconInsets


        let write = UIViewController()
        write.view.backgroundColor = UIColor.purple
        write.tabBarItem.image = UIImage(named: "fish")?.resizedImage(iconSize)
        write.tabBarItem.imageInsets = iconInsets

        let read = UIViewController()
        read.view.backgroundColor = UIColor.yellow
        read.tabBarItem.image = UIImage(named: "glasses")?.resizedImage(iconSize)
        read.tabBarItem.imageInsets = iconInsets

        let profile = UIViewController()
        profile.view.backgroundColor = UIColor.lightGray
        profile.tabBarItem.image = UIImage(named: "avatar")?.resizedImage(iconSize)
        profile.tabBarItem.imageInsets = iconInsets

        let tabBarController = UITabBarController()
        tabBarController.viewControllers = [home, story, read, profile]

        // Make the Tab Bar Controller the root view controller
        window?.rootViewController = tabBarController
        window?.makeKeyAndVisible()

        return true
    }
```

And StoryVC looks like this exactly like the Chat example project:
```
import Foundation
import MessageKit

class StoryVC: MessagesViewController {

    let dataSource = StoryDataSource()

    override func viewDidLoad() {
        super.viewDidLoad()

        messagesCollectionView.messagesDataSource = self
        messagesCollectionView.messageCellDelegate = self
        messagesCollectionView.messagesLayoutDelegate = self

        messageInputBar.delegate = self

        dataSource.delegate = self
    }
}

// MARK: - StoryDataSourceDelegate

extension StoryVC: StoryDataSourceDelegate {
    func storyDataUpdated() {
        messagesCollectionView.reloadData()
    }
}

// MARK: - MessagesDataSource

extension StoryVC: MessagesDataSource {

    func currentSender() -> Sender {
        return dataSource.hyou
    }

    func numberOfMessages(in messagesCollectionView: MessagesCollectionView) -> Int {
        return dataSource.msgs.count
    }

    func messageForItem(at indexPath: IndexPath, in messagesCollectionView: MessagesCollectionView) -> MessageType {
        return dataSource.msgs[indexPath.section]
    }
}

// MARK: - MessagesDisplayDataSource

extension StoryVC: MessagesDisplayDataSource {

    func avatar(for message: MessageType,
                at indexPath: IndexPath,
                in messagesCollectionView: MessagesCollectionView) -> Avatar
    {
        return dataSource.avatarFor(sender: message.sender)
    }

    func messageHeaderView(for message: MessageType,
                           at indexPath: IndexPath,
                           in messagesCollectionView: MessagesCollectionView)
        -> MessageHeaderView?
    {
        return messagesCollectionView.dequeueMessageHeaderView(for: indexPath)
    }

    func messageFooterView(for message: MessageType,
                           at indexPath: IndexPath,
                           in messagesCollectionView: MessagesCollectionView)
        -> MessageFooterView?
    {
        return messagesCollectionView.dequeueMessageFooterView(for: indexPath)
    }

    func cellTopLabelAttributedText(for message: MessageType,
                                    at indexPath: IndexPath)
        -> NSAttributedString?
    {
        let name = message.sender.displayName
        return NSAttributedString(
            string: name,
            attributes: [
                NSFontAttributeName: UIFont.preferredFont(forTextStyle: .caption1)
            ])
    }

    func cellBottomLabelAttributedText(for message: MessageType,
                                       at indexPath: IndexPath)
        -> NSAttributedString?
    {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        let dateString = formatter.string(from: message.sentDate)
        return NSAttributedString(
            string: dateString,
            attributes: [
                NSFontAttributeName: UIFont.preferredFont(forTextStyle: .caption2)
            ])
    }

}

// MARK: - MessagesLayoutDelegate

extension StoryVC: MessagesLayoutDelegate {

    func headerViewSize(for message: MessageType,
                        at indexPath: IndexPath,
                        in messagesCollectionView: MessagesCollectionView) -> CGSize
    {
        return CGSize(width: messagesCollectionView.bounds.width, height: 4)
    }

    func footerViewSize(for message: MessageType,
                        at indexPath: IndexPath,
                        in messagesCollectionView: MessagesCollectionView) -> CGSize
    {
        return CGSize(width: messagesCollectionView.bounds.width, height: 4)
    }
}

// MARK: - MessageCellDelegate

extension StoryVC: MessageCellDelegate {

    func didTapAvatar(in cell: MessageCollectionViewCell) {
        print("Avatar tapped")
    }

    func didTapMessage(in cell: MessageCollectionViewCell) {
        print("Message tapped")
    }

    func didTapBottomLabel(in cell: MessageCollectionViewCell) {}

    func didTapTopLabel(in cell: MessageCollectionViewCell) {}
}

// MARK: - MessageInputBarDelegate

extension StoryVC: MessageInputBarDelegate {

    func sendButtonPressed(sender: UIButton, textView: UITextView) {
        guard let text = textView.text else { return }
        dataSource.add(text, sender: currentSender())
    }
}
```

I did set break point and saw that the attributes.direction is set correctly and the cellTopLabel.textAlignment is set accordingly. Thoughts on what's causing the 2 issues?

**Comment by hyouuu** [comment]
BTW I found the current master's Example project has the 2nd issue as well - the name & date is always to the left.

![simulator screen shot aug 26 2017 1 33 06 pm](https://user-images.githubusercontent.com/5061867/29744905-1e2321c6-8a63-11e7-969b-9f7a09ee467c.png)

**Comment by hyouuu** [comment]
Regarding the 1st issue, it seems that's the general behavior on UIViewController's inputAccessoryView. Nevertheless it would be great if we could have a delegate method for when the keyboard is hidden, what offset we can set to the inputAccessoryView ;)

**Comment by SD10** [comment]
@hyouuu Thanks for reporting this. I'm pretty sure I know what the issue is regarding the label alignment. I can't believe I didn't notice it -- I was looking at that cell for hours yesterday lol.

About the tab bar controller. What if you set the `tabBar` to hidden? Usually, chat UI's don't display the tab bar as they want to take advantage of screen real estate. However, I'll still look into this.

**Comment by hyouuu** [comment]
Thanks for the quick reply! The reason to have the tabBar there is that user can switch among different tabs and check what's going on, and may not choose to actually start typing - with the tabBar hidden then the navigation goal cannot be achieved. I could change to push or present the view in if it's not easy to add a delegate method - looking forward to your findings!

**Comment by hyouuu** [comment]
Actually, even better than having a delegate method, if the inputAccessoryView could be attaching to the bottom of the message view it would solve all the problems automatically - currently I believe inputAccessoryView is attached to the UIWindow directly which is causing the issue

**Comment by SD10** [comment]
@hyouuu These should be fixed in `267dc55` and a few of the previous commits. Thanks for reporting this! 💯
